File: src/main/java/com/dansoftware/libraryapp/gui/util/WebsiteHyperLink.java
package com.dansoftware.libraryapp.gui.util;

import com.dansoftware.libraryapp.util.SystemBrowser;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.control.Hyperlink;
import javafx.scene.control.Tooltip;
import org.jetbrains.annotations.NotNull;

/**
 * A {@link WebsiteHyperLink} is a {@link Hyperlink} implementation that can
 * open websites in the default browser easily.
 *
 * <p>
 * It also has a {@link Tooltip} that shows the url.
 *
 * @author Daniel Gyorffy
 */
public class WebsiteHyperLink
        extends Hyperlink
        implements EventHandler<ActionEvent> {

    private final StringProperty url;

    public WebsiteHyperLink(@NotNull String text,
                            @NotNull String url) {
        this.url = new SimpleStringProperty(url);
        setTooltip(new HyperlinkTooltip(this));
        setText(text);
        setOnAction(this);
    }

    @Override
    public void handle(ActionEvent event) {
        if (SystemBrowser.isSupported()) {
            SystemBrowser systemBrowser = new SystemBrowser();
            systemBrowser.browse(url.get());
        }
    }

    public String getUrl() {
        return url.get();
    }

    public StringProperty urlProperty() {
        return url;
    }

    /**
     * The tooltip implementation
     */
    private static final class HyperlinkTooltip extends Tooltip {
        HyperlinkTooltip(WebsiteHyperLink parent) {
            textProperty().bind(parent.url);
        }
    }
}

File: src/main/java/com/dansoftware/libraryapp/gui/util/WebsiteHyperLink.kt
package com.dansoftware.libraryapp.gui.util

import com.dansoftware.libraryapp.util.SystemBrowser
import com.dansoftware.libraryapp.util.SystemBrowser.Companion.isSupported
import javafx.beans.property.SimpleStringProperty
import javafx.beans.property.StringProperty
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.Hyperlink
import javafx.scene.control.Tooltip

/**
 * A [WebsiteHyperLink] is a [Hyperlink] implementation that can
 * open websites in the default browser easily.
 *
 *
 *
 * It also has a [Tooltip] that shows the url.
 *
 * @author Daniel Gyorffy
 */
class WebsiteHyperLink(
    text: String,
    url: String
) : Hyperlink(), EventHandler<ActionEvent> {

    private val url: StringProperty

    init {
        this.url = SimpleStringProperty(url)
        this.tooltip = HyperlinkTooltip(this)
        this.text = text
        this.onAction = this
    }

    override fun handle(event: ActionEvent) {
        if (isSupported()) {
            val systemBrowser = SystemBrowser()
            systemBrowser.browse(url.get())
        }
    }

    fun getUrl(): String {
        return url.get()
    }

    fun urlProperty(): StringProperty = url

    /**
     * The tooltip implementation
     */
    private class HyperlinkTooltip(parent: WebsiteHyperLink) : Tooltip() {
        init {
            textProperty().bind(parent.url)
        }
    }
}